ably/hosts.go

```diff
@@ -0,0 +1,241 @@
+	defer restHosts.Unlock()
+	restHosts.visitedHosts = ablyutil.NewHashSet()
+}
+
+func (restHosts *restHosts) fallbackHostsRemaining() int {
+	restHosts.Lock()
+	defer restHosts.Unlock()
+	hosts, _ := restHosts.opts.getFallbackHosts()
+	return len(hosts) + 1 - len(restHosts.visitedHosts)
+}
+
+func (restHosts *restHosts) setPrimaryFallbackHost(host string) {
+	restHosts.Lock()
+	defer restHosts.Unlock()
+	restHosts.primaryFallbackHost = host
+}
+
+func (restHosts *restHosts) getPreferredHost() string {
+	restHosts.Lock()
+	defer restHosts.Unlock()
+	host := restHosts.cache.get()
+	if ablyutil.Empty(host) {
+		if ablyutil.Empty(restHosts.primaryFallbackHost) {
+			host = restHosts.getPrimaryHost()
+		} else {
+			host = restHosts.primaryFallbackHost
+		}
+	}
+	restHosts.visitedHosts.Add(host)
+	return host
+}
+
+func (restHosts *restHosts) cacheHost(host string) {
+	select {
+	case <-restHosts.cache.put(host):
+		return
+	case <-time.After(time.Second): // timeout of a second to cache the host
+		return
+	}
+}
+
+// RTN17
+type realtimeHosts struct {
+	opts         *clientOptions
+	visitedHosts ablyutil.HashSet
+	sync.Mutex
+}
+
+func newRealtimeHosts(opts *clientOptions) *realtimeHosts {
+	return &realtimeHosts{
+		opts:         opts,
+		visitedHosts: ablyutil.NewHashSet(),
+	}
+}
+
+func (realtimeHosts *realtimeHosts) getPrimaryHost() string {
+	return realtimeHosts.opts.getPrimaryRealtimeHost()
+}
+
+func (realtimeHosts *realtimeHosts) nextFallbackHost() string {
+	realtimeHosts.Lock()
+	defer realtimeHosts.Unlock()
+
+	getNonVisitedHost := func() string {
+		visitedHosts := realtimeHosts.visitedHosts
+		primaryHost := realtimeHosts.getPrimaryHost()
+		if !visitedHosts.Has(primaryHost) {
+			return primaryHost
+		}
+		hosts, _ := realtimeHosts.opts.getFallbackHosts()
+		shuffledFallbackHosts := ablyutil.Shuffle(hosts)
+		for _, host := range shuffledFallbackHosts {
+			if !visitedHosts.Has(host) {
+				return host
+			}
+		}
+		return ""
+	}
+
+	nonVisitedHost := getNonVisitedHost()
+	if !ablyutil.Empty(nonVisitedHost) {
+		realtimeHosts.visitedHosts.Add(nonVisitedHost)
+	}
+	return nonVisitedHost
+}
+
+func (realtimeHosts *realtimeHosts) resetVisitedFallbackHosts() {
+	realtimeHosts.Lock()
+	defer realtimeHosts.Unlock()
+	realtimeHosts.visitedHosts = ablyutil.NewHashSet()
+}
+
+func (realtimeHosts *realtimeHosts) fallbackHostsRemaining() int {
+	realtimeHosts.Lock()
+	defer realtimeHosts.Unlock()
+	hosts, _ := realtimeHosts.opts.getFallbackHosts()
+	return len(hosts) + 1 - len(realtimeHosts.visitedHosts)
+}
+
+func (realtimeHosts *realtimeHosts) getPreferredHost() string {
+	realtimeHosts.Lock()
+	defer realtimeHosts.Unlock()
+	host := realtimeHosts.getPrimaryHost() // primary host is always preferred host/ fallback host in realtime
+	realtimeHosts.visitedHosts.Add(host)
+	return host
+}
+
+// hostCache this caches a successful fallback host for 10 minutes.
+// Only used by REST client while making requests RSC15f
+type hostCache struct {
```

Comment:
This seems way too complicated, dealing unnecessary with concurrency and timeouts. How about something like:

```go
type hostCache struct {
    duration time.Duration

    sync.RWMutex
    deadline time.Time
    host string
}

func (c *hostCache) put(host string) {
    c.Lock()
    defer c.Unlock()
    c.host = host
    c.deadline = time.Now().Add(c.duration)
}

func (c *hostCache) get() string {
    c.RLock()
    defer c.RUnlock()
    if c.host == "" || time.Until(c.deadline) <= 0 {
        return ""
    }
    return c.host
}
```

Reply:
Yeah, this looks way more easy, actually earlier one was implemented by someone else, so I didn't try to change it

Reply:
Simplified the implementation !